File: python/algos/max_depth_bst.py
from typing import Optional


class TreeNode:
    def __init__(self, val=0, left=None, right=None):
        self.val = val
        self.left = left
        self.right = right


def max_depth_bst(tree: Optional[TreeNode]) -> bool:
    return 0 if tree is None else 1 + max(max_depth_bst(tree.left), max_depth_bst(tree.right))


if __name__ == '__main__':
    assert max_depth_bst(TreeNode(2)) == 3

File: rust/src/algos/max_depth_bst.rs
use std::cell::RefCell;
use std::cmp::max;
use std::rc::Rc;

#[derive(Debug, PartialEq, Eq)]
pub struct TreeNode {
    pub val: i32,
    pub left: Option<Rc<RefCell<TreeNode>>>,
    pub right: Option<Rc<RefCell<TreeNode>>>,
}

impl TreeNode {
    #[inline]
    pub fn new(val: i32) -> Self {
        TreeNode {
            val,
            left: None,
            right: None,
        }
    }
}

pub fn max_depth_bst(tree: Option<Rc<RefCell<TreeNode>>>) -> i32 {
    match tree {
        Some(node) => {
            1 + max(
                max_depth_bst(node.borrow().left.clone()),
                max_depth_bst(node.borrow().right.clone()),
            )
        }
        None => 0,
    }
}
